README.md
![SendGrid Logo](twilio_sendgrid_logo.png)

[![BuildStatus](https://github.com/sendgrid/smtpapi-go/actions/workflows/test-and-deploy.yml/badge.svg)](https://github.com/sendgrid/smtpapi-go/actions/workflows/test-and-deploy.yml)
[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](LICENSE)
[![Twitter Follow](https://img.shields.io/twitter/follow/sendgrid.svg?style=social&label=Follow)](https://twitter.com/sendgrid)
[![GitHub contributors](https://img.shields.io/github/contributors/sendgrid/smtpapi-go.svg)](https://github.com/sendgrid/smtpapi-go/graphs/contributors)
[![Go Report Card](https://goreportcard.com/badge/github.com/sendgrid/smtpapi-go)](https://goreportcard.com/report/github.com/sendgrid/smtpapi-go)
[![GoDoc](https://godoc.org/github.com/sendgrid/smtpapi-go?status.svg)](https://godoc.org/github.com/sendgrid/smtpapi-go)

**This is a simple library to simplify the process of using [SendGrid's](https://sendgrid.com) [X-SMTPAPI](http://sendgrid.com/docs/API_Reference/SMTP_API/index.html) with the Go programming language**

# Table of Contents

* [Announcements](#announcements)
* [Installation](#installation)
* [Quick Start](#quick-start)
* [Usage](#usage)
* [How to Contribute](#how-to-contribute)
* [About](#about)
* [License](#license)

# Announcements
**The default branch name for this repository has been changed to `main` as of 07/27/2020.**

All the updates to this library are documented in our [CHANGELOG](CHANGELOG.md).

<a name="installation"></a>
# Installation

## Supported Versions

* Go version 1.11-1.17

## Prerequisites

* The SendGrid service, starting at the [free level](https://sendgrid.com/free?source=smtpapi-go)

## Install the Package

```bash
go get github.com/sendgrid/smtpapi-go
```

## Setup the Environment Variables

### Environment Variable

Update the development environment with your [SENDGRID_API_KEY](https://app.sendgrid.com/settings/api_keys), for example:

```bash
echo "export SENDGRID_API_KEY='YOUR_API_KEY'" > sendgrid.env
echo "sendgrid.env" >> .gitignore
source ./sendgrid.env
```

<a name="quick-start"></a>
# Quick Start

```go
package main

import (
  "github.com/sendgrid/smtpapi-go"
  "fmt"
)

func main() {
  header := smtpapi.NewSMTPAPIHeader()
  header.AddTo("test@example.com")
  fmt.Println(header.JSONString())
}
```
<a name="usage"></a>
# Usage

* [SendGrid Docs](https://sendgrid.com/docs/API_Reference/SMTP_API/index.html)
* [Example Code](examples)

# How to Contribute

We encourage contribution to our libraries, please see our [CONTRIBUTING](CONTRIBUTING.md) guide for more details on contributions.

Quick links:

* [Feature Request](CONTRIBUTING.md#feature-request)
* [Bug Reports](CONTRIBUTING.md#submit-a-bug-report)
* [Improvements to the Codebase](CONTRIBUTING.md#improvements-to-the-codebase)
* [Review Pull Requests](CONTRIBUTING.md#code-reviews)

<a name="about"></a>
# About

smtpapi-go is maintained and funded by Twilio SendGrid, Inc. The names and logos for smtpapi-go are trademarks of Twilio SendGrid, Inc.

If you need help installing or using the library, please check the [Twilio SendGrid Support Help Center](https://support.sendgrid.com).

If you've instead found a bug in the library or would like new features added, go ahead and open issues or pull requests against this repo!

# License

[The MIT License (MIT)](LICENSE)


CHANGELOG.md
# Change Log
All the notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](http://semver.org/).

[2022-03-09] Version 0.6.9
--------------------------
**Library - Chore**
- [PR #91](https://github.com/sendgrid/smtpapi-go/pull/91): push Datadog Release Metric upon deploy success. Thanks to [@eshanholtz](https://github.com/eshanholtz)!


[2022-02-09] Version 0.6.8
--------------------------
**Library - Chore**
- [PR #90](https://github.com/sendgrid/smtpapi-go/pull/90): upgrade supported language versions. Thanks to [@childish-sambino](https://github.com/childish-sambino)!
- [PR #89](https://github.com/sendgrid/smtpapi-go/pull/89): add deploy job to test and deploy gh workflow. Thanks to [@Hunga1](https://github.com/Hunga1)!


[2021-12-15] Version 0.6.7
--------------------------
**Library - Chore**
- [PR #88](https://github.com/sendgrid/smtpapi-go/pull/88): migrate to GitHub Actions. Thanks to [@beebzz](https://github.com/beebzz)!


[2020-08-19] Version 0.6.6
--------------------------
**Library - Chore**
- [PR #86](https://github.com/sendgrid/smtpapi-go/pull/86): update GitHub branch references. Thanks to [@thinkingserious](https://github.com/thinkingserious)!

**Library - Docs**
- [PR #68](https://github.com/sendgrid/smtpapi-go/pull/68): Run *.md documents through Grammer.ly. Thanks to [@ssiddhantsharma](https://github.com/ssiddhantsharma)!


[2020-08-05] Version 0.6.5
--------------------------
**Library - Test**
- [PR #56](https://github.com/sendgrid/smtpapi-go/pull/56): add CodeCov support to .travis.yml. Thanks to [@mptap](https://github.com/mptap)!


[2020-07-08] Version 0.6.4
--------------------------
**Library - Docs**
- [PR #82](https://github.com/sendgrid/smtpapi-go/pull/82): add use cases directory. Thanks to [@mtroiani](https://github.com/mtroiani)!


[2020-02-19] Version 0.6.3
--------------------------
**Library - Docs**
- [PR #61](https://github.com/sendgrid/smtpapi-go/pull/61): Added Code Review to Contributing.md. Thanks to [@mptap](https://github.com/mptap)!


[2020-02-05] Version 0.6.2
--------------------------
**Library - Docs**
- [PR #77](https://github.com/sendgrid/smtpapi-go/pull/77): Fixed link to bug report template. Thanks to [@alxshelepenok](https://github.com/alxshelepenok)!


[2020-01-30] Version 0.6.1
--------------------------
**Library - Docs**
- [PR #85](https://github.com/sendgrid/smtpapi-go/pull/85): baseline all the templated markdown docs. Thanks to [@childish-sambino](https://github.com/childish-sambino)!
- [PR #80](https://github.com/sendgrid/smtpapi-go/pull/80): added announcement to README. Thanks to [@luciajimenez](https://github.com/luciajimenez)!

**Library - Chore**
- [PR #84](https://github.com/sendgrid/smtpapi-go/pull/84): prep repo for automation. Thanks to [@thinkingserious](https://github.com/thinkingserious)!
- [PR #65](https://github.com/sendgrid/smtpapi-go/pull/65): fix failing build. Thanks to [@vaskoz](https://github.com/vaskoz)!
- [PR #69](https://github.com/sendgrid/smtpapi-go/pull/69): add style standards check. Thanks to [@pangaunn](https://github.com/pangaunn)!


[2018-06-04] Version 0.6.0
--------------------------
### Added
- PR [#24](https://github.com/sendgrid/smtpapi-go/pull/24): Added Code of Conduct. Big thanks to [Shivam Agarwal](https://github.com/gr8shivam) for the PR!
- PR [#26](https://github.com/sendgrid/smtpapi-go/pull/26): Added TROUBLESHOOTING.md. Big thanks to [Callam Delaney](https://github.com/cal97g) for the PR!
- PR [#30](https://github.com/sendgrid/smtpapi-go/pull/30): SEO updates. Big thanks to [Eric Liu](https://github.com/eric1iu) for the PR!
- PR [#32](https://github.com/sendgrid/smtpapi-go/pull/32): Added table of contents to README.md. Big thanks to [thepriefy](https://github.com/thepriefy) for the PR!
- PR [#29](https://github.com/sendgrid/smtpapi-go/pull/29): Added README.md updates. Big thanks to [Andrew Hamilton](https://github.com/ahamilton55) for the PR!
- PR [#36](https://github.com/sendgrid/smtpapi-go/pull/36): Added .github/ISSUE_TEMPLATE. Big thanks to [thepriefy](https://github.com/thepriefy) for the PR!
- PR [#31](https://github.com/sendgrid/smtpapi-go/pull/31): Parallelize all tests. Big thanks to [Vasko Zdravevski](https://github.com/vaskoz) for the PR!
- PR [#38](https://github.com/sendgrid/smtpapi-go/pull/38): Created PULL_REQUEST_TEMPLATE. Big thanks to [Anatoly](https://github.com/anatolyyyyyy) for the PR!
- PR [#51](https://github.com/sendgrid/smtpapi-go/pull/51): Don't ignore erros in test. Big thanks to [Revolta](https://github.com/keydrevolta) for the PR!
- PR [#52](https://github.com/sendgrid/smtpapi-go/pull/52): Include Gometalinter in the Travis CI build. Big thanks to [Vasko Zdravevski](https://github.com/vaskoz) for the PR!
- PR [#42](https://github.com/sendgrid/smtpapi-go/pull/42): Added a .env_sample file, updated gitignore, updated README.md. Big thanks to [thepriefy](https://github.com/thepriefy) for the PR!
- PR [#44](https://github.com/sendgrid/smtpapi-go/pull/44): Added Usage.md. Big thanks to [Tariq Ibrahim](https://github.com/tariq1890) for the PR!
- PR [#49](https://github.com/sendgrid/smtpapi-go/pull/49): Added tests(License): The end year in the license file should be current year. Big thanks to [Marco Antônio Singer](https://github.com/marcosinger) for the PR!
- PR [#53](https://github.com/sendgrid/smtpapi-go/pull/53): Added tests to check for the certain repo files to exist in the repo. Big thanks to [Gabriel Lima](https://github.com/gabrielclima) for the PR!
- PR [#63](https://github.com/sendgrid/smtpapi-go/pull/63): Adding dynamic data support. Big thanks to [Ron Canada](https://github.com/roncanada) for the PR!

### Fixed
- PR [#21](https://github.com/sendgrid/smtpapi-go/pull/21): README.md typo. Big thanks to [Cícero Pablo](https://github.com/ciceropablo) for the PR!

smtpapi_test.go
package smtpapi

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"reflect"
	"regexp"
	"testing"
	"time"
)

func exampleJson() map[string]interface{} {
	data, _ := ioutil.ReadFile("smtpapi_test_strings.json")
	var f interface{}
	json.Unmarshal(data, &f)
	json := f.(map[string]interface{})
	return json
}

func TestNewSMTPIAPIHeader(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	if header == nil {
		t.Error("NewSMTPAPIHeader() should never return nil")
	}
}

func TestAddTo(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	header.AddTo("addTo@mailinator.com")
	result, _ := header.JSONString()
	if result != exampleJson()["add_to"] {
		t.Errorf("Result did not match")
	}
}

func TestAddTos(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	tos := []string{"addTo@mailinator.com"}
	header.AddTos(tos)
	result, _ := header.JSONString()
	if result != exampleJson()["add_to"] {
		t.Errorf("Result did not match")
	}
}

func TestSetTos(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	header.SetTos([]string{"setTos@mailinator.com"})
	result, _ := header.JSONString()
	if result != exampleJson()["set_tos"] {
		t.Errorf("Result did not match")
	}
}

func TestAddSubstitution(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	header.AddSubstitution("sub", "val")
	result, _ := header.JSONString()
	if result != exampleJson()["add_substitution"] {
		t.Errorf("Result did not match")
	}
}

func TestAddSubstitutions(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	header.AddSubstitutions("sub", []string{"val"})
	result, _ := header.JSONString()
	if result != exampleJson()["add_substitution"] {
		t.Errorf("Result did not match")
	}
}

func TestSetSubstitutions(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	sub := make(map[string][]string)
	sub["sub"] = []string{"val"}
	header.SetSubstitutions(sub)
	result, _ := header.JSONString()
	if result != exampleJson()["set_substitutions"] {
		t.Errorf("Result did not match")
	}
}

func TestAddSection(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	header.AddSection("set_section_key", "set_section_value")
	header.AddSection("set_section_key_2", "set_section_value_2")
	result, _ := header.JSONString()
	if result != exampleJson()["add_section"] {
		t.Errorf("Result did not match")
	}
}

func TestSetSections(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	sections := make(map[string]string)
	sections["set_section_key"] = "set_section_value"
	header.SetSections(sections)
	result, _ := header.JSONString()
	if result != exampleJson()["set_sections"] {
		t.Errorf("Result did not match")
	}
}

func TestAddCategory(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	header.AddCategory("addCategory")
	header.AddCategory("addCategory2")
	result, _ := header.JSONString()
	if result != exampleJson()["add_category"] {
		t.Errorf("Result did not match")
	}
}

func TestAddCategoryUnicode(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	header.AddCategory("カテゴリUnicode")
	header.AddCategory("カテゴリ2Unicode")
	header.AddCategory("鼖")
	result, _ := header.JSONString()
	if result != exampleJson()["add_category_unicode"] {
		t.Errorf("Result did not match")
	}
}

func TestAddCategories(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	categories := []string{"addCategory", "addCategory2"}
	header.AddCategories(categories)
	result, _ := header.JSONString()
	if result != exampleJson()["add_category"] {
		t.Errorf("Result did not match")
	}
}

func TestSetCategories(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	header.SetCategories([]string{"setCategories"})
	result, _ := header.JSONString()
	if result != exampleJson()["set_categories"] {
		t.Errorf("Result did not match")
	}
}

func TestAddUniqueArg(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	header.AddUniqueArg("add_unique_argument_key", "add_unique_argument_value")
	header.AddUniqueArg("add_unique_argument_key_2", "add_unique_argument_value_2")
	result, _ := header.JSONString()
	if result != exampleJson()["add_unique_arg"] {
		t.Errorf("Result did not match")
	}
}

func TestSetUniqueArgs(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	args := make(map[string]string)
	args["set_unique_argument_key"] = "set_unique_argument_value"
	header.SetUniqueArgs(args)
	result, _ := header.JSONString()
	if result != exampleJson()["set_unique_args"] {
		t.Errorf("Result did not match")
	}
}

func TestAddFilter(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	header.AddFilter("footer", "text/html", "<strong>boo</strong>")
	if len(header.Filters) != 1 {
		t.Error("AddFilter failed")
	}
}

func TestSetFilter(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	filter := &Filter{
		Settings: make(map[string]interface{}),
	}
	filter.Settings["enable"] = 1
	filter.Settings["text/plain"] = "You can haz footers!"
	header.SetFilter("footer", filter)
	result, _ := header.JSONString()
	if result != exampleJson()["set_filters"] {
		t.Errorf("Result did not match")
	}
}

func TestSetSendAt(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	header.SetSendAt(1428611024)
	result, _ := header.JSONString()
	if result != exampleJson()["set_send_at"] {
		t.Errorf("Result did not match")
	}
}

func TestAddSendEachAt(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	header.AddSendEachAt(1428611024)
	header.AddSendEachAt(1428611025)
	result, _ := header.JSONString()
	if result != exampleJson()["add_send_each_at"] {
		t.Errorf("Result did not match")
	}
}

func TestSetSendEachAt(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	sendEachAt := []int64{1428611024, 1428611025}
	header.SetSendEachAt(sendEachAt)
	result, _ := header.JSONString()
	if result != exampleJson()["set_send_each_at"] {
		t.Errorf("Result did not match")
	}
}

func TestSetASMGroupID(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	header.SetASMGroupID(1)
	result, _ := header.JSONString()
	if result != exampleJson()["set_asm_group_id"] {
		t.Errorf("Result did not match")
	}
}

func TestSetIpPool(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	header.SetIpPool("testPool")
	result, _ := header.JSONString()
	if result != exampleJson()["set_ip_pool"] {
		t.Errorf("Result did not match")
	}
}

func TestSAddASMGroupToDisplay(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	header.AddASMGroupToDisplay(671332)
	result, _ := header.JSONString()
	if result != exampleJson()["add_asm_group"] {
		t.Errorf("Result did not match")
	}
}

func TestSAddASMGroupsToDisplay(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	header.AddASMGroupsToDisplay([]int{45, 23})
	result, _ := header.JSONString()
	if result != exampleJson()["add_asm_groups"] {
		t.Errorf("Result did not match")
	}
}

func TestJSONString(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	result, _ := header.JSONString()
	if result != exampleJson()["json_string"] {
		t.Errorf("Result did not match")
	}
}

func TestJSONStringWithAdds(t *testing.T) {
	t.Parallel()
	validHeader, _ := json.Marshal([]byte(`{"to":["test@email.com"],"sub":{"subKey":["subValue"]},"section":{"testSection":"sectionValue"},"category":["testCategory"],"unique_args":{"testUnique":"uniqueValue"},"filters":{"testFilter":{"settings":{"filter":"filterValue"}}}}`))
	header := NewSMTPAPIHeader()
	header.AddTo("test@email.com")
	header.AddSubstitution("subKey", "subValue")
	header.AddSection("testSection", "sectionValue")
	header.AddCategory("testCategory")
	header.AddUniqueArg("testUnique", "uniqueValue")
	header.AddFilter("testFilter", "filter", "filterValue")
	if h, e := header.JSONString(); e != nil {
		t.Errorf("Error! %s", e)
	} else {
		testHeader, _ := json.Marshal([]byte(h))
		if reflect.DeepEqual(testHeader, validHeader) {
			t.Logf("Success")
		} else {
			t.Errorf("Invalid header")
		}
	}
}

func TestJSONStringWithSets(t *testing.T) {
	t.Parallel()
	validHeader, _ := json.Marshal([]byte(`{"to":["test@email.com"],"sub":{"subKey":["subValue"]},"section":{"testSection":"sectionValue"},"category":["testCategory"],"unique_args":{"testUnique":"uniqueValue"},"filters":{"testFilter":{"settings":{"filter":"filterValue"}}},"asm_group_id":1,"ip_pool":"testPool"}`))
	header := NewSMTPAPIHeader()
	header.SetTos([]string{"test@email.com"})
	sub := make(map[string][]string)
	sub["subKey"] = []string{"subValue"}
	header.SetSubstitutions(sub)
	sections := make(map[string]string)
	sections["testSection"] = "sectionValue"
	header.SetSections(sections)
	header.SetCategories([]string{"testCategory"})
	unique := make(map[string]string)
	unique["testUnique"] = "uniqueValue"
	header.SetUniqueArgs(unique)
	header.AddFilter("testFilter", "filter", "filterValue")
	header.SetASMGroupID(1)
	header.SetIpPool("testPool")
	if h, e := header.JSONString(); e != nil {
		t.Errorf("Error! %s", e)
	} else {
		testHeader, _ := json.Marshal([]byte(h))
		if reflect.DeepEqual(testHeader, validHeader) {
			t.Logf("Success")
		} else {
			t.Errorf("Invalid header")
		}
	}
}

func TestMarshalUnmarshall(t *testing.T) {
	t.Parallel()
	header := NewSMTPAPIHeader()
	header.SetTos([]string{"test@email.com"})
	sub := make(map[string][]string)
	sub["subKey"] = []string{"subValue"}
	header.SetSubstitutions(sub)
	sections := make(map[string]string)
	sections["testSection"] = "sectionValue"
	header.SetSections(sections)
	header.SetCategories([]string{"testCategory"})
	unique := make(map[string]string)
	unique["testUnique"] = "uniqueValue"
	header.SetUniqueArgs(unique)
	header.AddFilter("testFilter", "filter", "filterValue")
	header.SetASMGroupID(1)
	header.SetIpPool("testPool")
	header.SetASMGroupsToDisplay([]int{32, 12})

	newHeader := NewSMTPAPIHeader()
	b, err := header.JSONString()
	if err != nil {
		t.Errorf("Error in JSONString %v", err)
	}
	err = newHeader.Load([]byte(b))
	if err != nil {
		t.Errorf("Could not load newHeader %v", err)
	}
	if !reflect.DeepEqual(header, newHeader) {
		t.Errorf("Expected %v, but got %v", header, newHeader)
	}
}

func TestRepoFiles(t *testing.T) {
	files := []string{".env_sample", ".gitignore", ".github/workflows/test-and-deploy.yml", "CHANGELOG.md", "CODE_OF_CONDUCT.md",
		"CONTRIBUTING.md", "ISSUE_TEMPLATE.md", "LICENSE", "PULL_REQUEST_TEMPLATE.md",
		"README.md", "TROUBLESHOOTING.md", "USAGE.md"}

	for _, file := range files {
		if _, err := os.Stat(file); os.IsNotExist(err) {
			t.Errorf("Repo file does not exist: %v", file)
		}
	}
}

func TestLicenseYear(t *testing.T) {
	t.Parallel()
	dat, err := ioutil.ReadFile("LICENSE")

	currentYear := time.Now().Year()
	r := fmt.Sprintf("%d", currentYear)
	match, _ := regexp.MatchString(r, string(dat))

	if err != nil {
		t.Error("License File Not Found")
	}
	if !match {
		t.Error("Incorrect Year in License Copyright")
	}
}


smtpapi.go
package smtpapi

import (
	"bytes"
	"encoding/json"
	"fmt"
	"unicode/utf16"
)

// Version represents the current version of the smtpapi-go library
const Version = "0.6.9"

// SMTPAPIHeader will be used to set up X-SMTPAPI params
type SMTPAPIHeader struct {
	To          []string               `json:"to,omitempty"`
	Sub         map[string][]string    `json:"sub,omitempty"`
	Section     map[string]string      `json:"section,omitempty"`
	Category    []string               `json:"category,omitempty"`
	UniqueArgs  map[string]string      `json:"unique_args,omitempty"`
	Filters     map[string]Filter      `json:"filters,omitempty"`
	ASMGroupID  int                    `json:"asm_group_id,omitempty"`
	ASMGroups   []int                  `json:"asm_groups_to_display,omitempty"`
	SendAt      int64                  `json:"send_at,omitempty"`
	SendEachAt  []int64                `json:"send_each_at,omitempty"`
	IpPool      string                 `json:"ip_pool,omitempty"`
	BatchID     string                 `json:"batch_id,omitempty"`
	DynamicData map[string]interface{} `json:"dynamic_template_data,omitempty"`
}

// Filter represents an App/Filter and its settings
type Filter struct {
	Settings map[string]interface{} `json:"settings,omitempty"`
}

// NewSMTPAPIHeader creates a new header struct
func NewSMTPAPIHeader() *SMTPAPIHeader {
	return &SMTPAPIHeader{}
}

// AddTo appends a single email to the To header
func (h *SMTPAPIHeader) AddTo(email string) {
	h.To = append(h.To, email)
}

// AddTos appends multiple emails to the To header
func (h *SMTPAPIHeader) AddTos(emails []string) {
	for i := 0; i < len(emails); i++ {
		h.AddTo(emails[i])
	}
}

// SetTos sets the value of the To header
func (h *SMTPAPIHeader) SetTos(emails []string) {
	h.To = emails
}

// AddSubstitution adds a new substitution to a specific key
func (h *SMTPAPIHeader) AddSubstitution(key, sub string) {
	if h.Sub == nil {
		h.Sub = make(map[string][]string)
	}
	h.Sub[key] = append(h.Sub[key], sub)
}

// AddSubstitutions adds a multiple substitutions to a specific key
func (h *SMTPAPIHeader) AddSubstitutions(key string, subs []string) {
	for i := 0; i < len(subs); i++ {
		h.AddSubstitution(key, subs[i])
	}
}

// SetSubstitutions sets the value of the substitutions on the Sub header
func (h *SMTPAPIHeader) SetSubstitutions(sub map[string][]string) {
	h.Sub = sub
}

// AddSection sets the value for a specific section
func (h *SMTPAPIHeader) AddSection(section, value string) {
	if h.Section == nil {
		h.Section = make(map[string]string)
	}
	h.Section[section] = value
}

// SetSections sets the value for the Section header
func (h *SMTPAPIHeader) SetSections(sections map[string]string) {
	h.Section = sections
}

// AddCategory adds a new category to the Category header
func (h *SMTPAPIHeader) AddCategory(category string) {
	h.Category = append(h.Category, category)
}

// AddCategories adds multiple categories to the Category header
func (h *SMTPAPIHeader) AddCategories(categories []string) {
	for i := 0; i < len(categories); i++ {
		h.AddCategory(categories[i])
	}
}

// SetCategories will set the value of the Categories field
func (h *SMTPAPIHeader) SetCategories(categories []string) {
	h.Category = categories
}

// SetASMGroupID will set the value of the ASMGroupID field
func (h *SMTPAPIHeader) SetASMGroupID(groupID int) {
	h.ASMGroupID = groupID
}

// AddASMGroupToDisplay adds a new ASM group ID to be displayed
func (h *SMTPAPIHeader) AddASMGroupToDisplay(groupID int) {
	h.ASMGroups = append(h.ASMGroups, groupID)
}

// AddASMGroupsToDisplay adds multiple ASM group IDs to be displayed
func (h *SMTPAPIHeader) AddASMGroupsToDisplay(groupIDs []int) {
	for i := 0; i < len(groupIDs); i++ {
		h.AddASMGroupToDisplay(groupIDs[i])
	}
}

// SetASMGroupsToDisplay will set the value of the ASMGroups field
func (h *SMTPAPIHeader) SetASMGroupsToDisplay(groups []int) {
	h.ASMGroups = groups
}

// AddUniqueArg will set the value of a specific argument
func (h *SMTPAPIHeader) AddUniqueArg(arg, value string) {
	if h.UniqueArgs == nil {
		h.UniqueArgs = make(map[string]string)
	}
	h.UniqueArgs[arg] = value
}

// SetUniqueArgs will set the value of the Unique_args header
func (h *SMTPAPIHeader) SetUniqueArgs(args map[string]string) {
	h.UniqueArgs = args
}

// AddFilter will set the specific setting for a filter
func (h *SMTPAPIHeader) AddFilter(filter, setting string, value interface{}) {
	if h.Filters == nil {
		h.Filters = make(map[string]Filter)
	}
	if _, ok := h.Filters[filter]; !ok {
		h.Filters[filter] = Filter{
			Settings: make(map[string]interface{}),
		}
	}
	h.Filters[filter].Settings[setting] = value
}

// SetFilter takes in a Filter struct with predetermined settings and sets it for such Filter key
func (h *SMTPAPIHeader) SetFilter(filter string, value *Filter) {
	if h.Filters == nil {
		h.Filters = make(map[string]Filter)
	}
	h.Filters[filter] = *value
}

// SetSendAt takes in a timestamp which determines when the email will be sent
func (h *SMTPAPIHeader) SetSendAt(sendAt int64) {
	h.SendAt = sendAt
}

// AddSendEachAt takes in a timestamp and pushes it into a list Must match length of To emails
func (h *SMTPAPIHeader) AddSendEachAt(sendEachAt int64) {
	h.SendEachAt = append(h.SendEachAt, sendEachAt)
}

// SetSendEachAt takes an array of timestamps. Must match length of To emails
func (h *SMTPAPIHeader) SetSendEachAt(sendEachAt []int64) {
	h.SendEachAt = sendEachAt
}

// SetIpPool takes a strings and sets the IpPool field
func (h *SMTPAPIHeader) SetIpPool(ipPool string) {
	h.IpPool = ipPool
}

// Unicode escape
func escapeUnicode(input string) string {
	//var buffer bytes.Buffer
	buffer := bytes.NewBufferString("")
	for _, r := range input {
		if r > 65535 {
			// surrogate pair
			var r1, r2 = utf16.EncodeRune(r)
			var s = fmt.Sprintf("\\u%x\\u%x", r1, r2)
			// error always nil https://golang.org/pkg/bytes/#Buffer.WriteString
			buffer.WriteString(s) // nolint: gas, gosec
		} else if r > 127 {
			var s = fmt.Sprintf("\\u%04x", r)
			// error always nil https://golang.org/pkg/bytes/#Buffer.WriteString
			buffer.WriteString(s) // nolint: gas, gosec
		} else {
			var s = fmt.Sprintf("%c", r)
			// error always nil https://golang.org/pkg/bytes/#Buffer.WriteString
			buffer.WriteString(s) // nolint: gas, gosec
		}
	}
	return buffer.String()
}

// JSONString returns the representation of the Header
func (h *SMTPAPIHeader) JSONString() (string, error) {
	headers, e := json.Marshal(h)
	return escapeUnicode(string(headers)), e
}

// Load allows you to load a pre-formed x-smtpapi header
func (h *SMTPAPIHeader) Load(b []byte) error {
	return json.Unmarshal(b, h)
}


examples/example.go
package main

import (
	"fmt"
	"github.com/sendgrid/smtpapi-go"
)

func main() {
	header := smtpapi.NewSMTPAPIHeader()

	// Recipients
	header.AddTo("test@example.com")
	// or
	//tos := []string{"test1@example.com", "test2@example.com"}
	//header.AddTos(tos)
	// or
	//header.SetTos(tos)

	//[Substitutions](http://sendgrid.com/docs/API_Reference/SMTP_API/substitution_tags.html)
	header.AddSubstitution("key", "value")
	// or
	//values := []string{"value1", "value2"}
	//header.AddSubstitutions("key", values)
	// or
	//sub := make(map[string][]string)
	//sub["key"] = values
	//header.SetSubstitutions(sub)

	//[Section](http://sendgrid.com/docs/API_Reference/SMTP_API/section_tags.html)
	header.AddSection("section", "value")
	// or
	//sections := make(map[string]string)
	//sections["section"] = "value"
	//header.SetSections(sections)

	//[Category](http://sendgrid.com/docs/Delivery_Metrics/categories.html)
	header.AddCategory("category")
	// or
	//categories := []string{"setCategories"}
	//header.AddCategories(categories)
	// or
	//header.SetCategories(categories)

	//[Unique Arguments](http://sendgrid.com/docs/API_Reference/SMTP_API/unique_arguments.html)
	header.AddUniqueArg("key", "value")
	// or
	//args := make(map[string]string)
	//args["key"] = "value"
	//header.SetUniqueArgs(args)

	//[Filters](http://sendgrid.com/docs/API_Reference/SMTP_API/apps.html)
	header.AddFilter("filter", "setting", "value")
	// or
	/*
	  filter := &Filter{
	    Settings: make(map[string]string),
	  }
	  filter.Settings["enable"] = "1"
	  filter.Settings["text/plain"] = "You can haz footers!"
	  header.SetFilter("footer", filter)
	*/

	//[Send At](https://sendgrid.com/docs/API_Reference/SMTP_API/scheduling_parameters.html)
	header.SetSendAt(1428611024)
	// or
	//sendEachAt := []int64{1428611024, 1428611025}
	//header.SetSendEachAt(sendEachAt)
	// or
	//header.AddSendEachAt(1428611024)
	//header.AddSendEachAt(1428611025)

	//[ASM Group ID](https://sendgrid.com/docs/User_Guide/advanced_suppression_manager.html)
	asmGroupID := 1
	header.SetASMGroupID(asmGroupID)

	// [IP Pools](https://sendgrid.com/docs/API_Reference/Web_API_v3/IP_Management/ip_pools.html)
	header.SetIpPool("testPool")

	fmt.Println(header.JSONString())
}
